library/init/meta/case_tag.lean

```diff
@@ -0,0 +1,250 @@
+   which hypotheses to rename; this information is contained in the case tag for
+   the `cons` goal.
+
+## Module contents
+
+This module defines
+
+1. what a case tag is (see `case_tag`);
+2. how to render a `case_tag` as a list of names (see `render`);
+3. how to parse  a `case_tag` from a list of names (see `parse`);
+4. how to match a `case_tag` with a sequence of names given by the user
+   (see `match_tag`).
+-/
+
+namespace tactic
+namespace interactive
+
+/--
+A case tag carries the following information:
+
+1. A list of names identifying the case (`names`). This is a list of constructor
+   names, one for each case split that was performed. For example, the sequence
+   of tactics `cases n; cases xs`, where `n` is a natural number and `xs` is a
+   list, will generate four cases tagged as follows:
+
+   ```
+   nat.zero, list.nil
+   nat.zero, list.cons
+   nat.succ, list.nil
+   nat.succ, list.cons
+   ```
+
+   Note: In the case tag, these names are stored in reverse order. Thus, the
+   `names` of the first case tag would be `list.nil, nat.zero`.
+
+2. Information about the arguments introduced by the cases-like tactic.
+   Different tactics work slightly different in this regard:
+
+   1. The `with_cases` tactic generates goals where the target quantifies over
+      any added hypotheses. For example, `with_cases { cases xs }`, where `xs`
+      is a `list α`, will generate a target of the form `α → list α → ...` in
+      the `cons` case, where the two arguments correspond to the two arguments
+      of the `cons` constructor. Goals of this form are tagged with a `pi` case
+      tag (since the target is a pi type). In addition to the `names`, it
+      contains a natural number, `num_arguments`, which specifies how many of
+      the arguments that the target quantifies over were introduced by
+      `with_cases`.
+   2. The `cases` and `induction` tactics do not add arguments to the target,
+      but rather introduce them as hypotheses in the local context. Goals of
+      this form are tagged with a `hyps` case tag. In addition to the `names`,
+      it contains a list of *unique* names of the hypotheses that were
+      introduced.
+
+      Note: Many tactics do not preserve the unique names of hypotheses
+      (particularly those tactics that use `revert`). Therefore, a `hyps` case
+      tag is only guaranteed to be valid directly after it was generated.
+-/
+inductive case_tag
+| pi (names : list name) (num_arguments : ℕ)
+| hyps (names : list name) (arguments : list name)
+
+/--
+Display a case tag.
+-/
+protected def case_tag.to_string : case_tag → string
+| (case_tag.pi names num_arguments) :=
+  "(case_tag.pi " ++ to_string names ++ " " ++ to_string num_arguments ++ ")"
+| (case_tag.hyps names arguments) :=
+  "(case_tag.hyps " ++ to_string names ++ " " ++ to_string arguments ++ ")"
+
+namespace case_tag
+
+open name (mk_string mk_numeral)
+
+instance : has_to_string case_tag :=
```

Comment:
There seems to be no instance `has_repr name` (which blocks `has_repr case_tag`). Is this intentional?

Reply:
I think the *intention* behind the `repr`/`to_string` split is that `repr` should output "valid" Lean code, and `to_string` a "Human-readable" version.  But we are missing lots of instances because of this stringification class inflation.  Practically speaking, it's really frustrating even if only some of the instances are missing because you need all of them.  For example, `#eval` want `has_repr`, `pp` wants `has_to_tactic_format`, etc.

So yes, please add `has_repr name`.  And add `has_repr case_tag` no matter whether there is a `has_repr name` or not.  (I would have been ok with `instance : has_repr case_tag := ⟨to_string⟩`.)

Reply:
I get the `repr`/`to_string` split; there's a semantic difference there. What I think is unnecessary is `has_format`/`to_string`. If `format` wasn't `meta` -- and I don't see why it should be -- `to_string` could be subsumed by `has_format`.

Anyway, I'll add the `has_repr` instances.

Reply:
Oh, I hit the merge button a bit too early.

BTW, `to_string` is *not* subsumed by `to_fmt`.  There is no `has_to_string` instance using `has_to_format`.

Reply:
Yes, my point was that `to_string` *could* be subsumed by `to_fmt` if `has_format` wasn't `meta`, because `to_format` and `to_string` serve the same purpose (human-friendly printing).

Reply:
:+1: for adding a low-priority instance converting from `has_to_format` to `has_to_string`.